### PR TITLE
fixup stable balancer, added generalized weighted

### PR DESCRIPTION
**Pull request by ksyao2002**
Only the stable balancer pairs are tested. I haven't tested the new weighted formula yet, but it was sourced from midas' code: https://github.com/Midas-Protocol/contracts/blob/352be0e9ba2795e14d05a5fa4661cb2569655141/contracts/oracles/default/BalancerLpTokenPriceOracleNTokens.sol. It should be very similar, I just changed the decimals so that it is compatible with the 8 decimal system we use on optimism

I will test the weighted formula later using mocked tokens.

### CODE REVIEW

File: packages/contracts/contracts/protocol/oracles/BalancerOracle.sol [hunk 1]
```diff
@@ -57,124 +63,189 @@ library BalancerOracle {
...
-  	    fairResA = BNum.bdiv(resA, BNum.bpow(ratio, wB));
-  	    fairResB = BNum.bmul(resB, BNum.bpow(ratio, wA));
-  	  }
-	  //check for low pool ratios
-	  require(ratio > 1e12, "reserves too low"); 
-  	}
+	/// @param reserves Reserves of the assets
+	/// @param weights Weights of the assets
+	/// @param prices Fair prices of the assets
+	function computeFairReserves(
```
Comment (fico23) [inline packages/contracts/contracts/protocol/oracles/BalancerOracle.sol:71]: I dont understand what is happening here so I'll just comment on the general code guidelines:

switch all for loops to

`for (uint256 i; i < length;) {
unchecked{++i}
}`

cache lengths: uint256 length = reserves.length

also all of these variables could be initialized before first for loop, and then reused in every iteration

File: packages/contracts/localhost_tests_OP/test-oracle.spec.ts
Comment (fico23) [inline packages/contracts/localhost_tests_OP/test-oracle.spec.ts:1]: We should have atleast one test that first execute a huge buy, which makes balancer pool "unbalanced", for example having 99% assets in USDT and 1% in USDC/DAI. then see how this price oracle behaves

**Review by mel0ndev — APPROVED** [review]

File: packages/contracts/contracts/protocol/oracles/BalancerOracle.sol (hunk 1 above)
Comment (ksyao2002) [reply packages/contracts/contracts/protocol/oracles/BalancerOracle.sol:71]: I copied these formulas directly from midas. I think if we want to avoid another audit, we should keep the code exactly the same as the audited code from midas. Maybe we can include these in the next audit?

Comment (fico23) [reply packages/contracts/contracts/protocol/oracles/BalancerOracle.sol:71]: Ok, sounds good. These nested for loops look suspicious though.
We should still include more tests in edge cases where pools become unbalanced

File: packages/contracts/localhost_tests_OP/test-oracle.spec.ts
Comment (ksyao2002) [reply packages/contracts/localhost_tests_OP/test-oracle.spec.ts:1]: Finished testing everything. Found out that the weighted formula is actually manipulatable through swaps. removed for now.

File: packages/contracts/contracts/protocol/oracles/BalancerOracle.sol (hunk 1 above)
Comment (ksyao2002) [reply packages/contracts/contracts/protocol/oracles/BalancerOracle.sol:74]: I removed this
